### PR TITLE
qsvvpp: add support for HDR-to-SDR tone-mapping

### PR DESCRIPTION
**Pull request by xhaihao**

### CODE REVIEW

File: libavcodec/qsvdec.c [hunk 1]
```diff
@@ -628,6 +645,54 @@ static int qsv_export_film_grain(AVCodecContext *avctx, mfxExtAV1FilmGrainParam
...
+#if QSV_VERSION_ATLEAST(1, 35)
+static int qsv_export_hdr_side_data(AVCodecContext *avctx, mfxExtMasteringDisplayColourVolume *mdcv,
+                                    mfxExtContentLightLevelInfo *clli, AVFrame *frame)
+{
+    // The SDK re-uses this flag for HDR SEI parsing
+    if (mdcv->InsertPayloadToggle) {
+        AVMasteringDisplayMetadata *mastering = av_mastering_display_metadata_create_side_data(frame);
+        const int mapping[3] = {2, 0, 1};
+        const int chroma_den = 50000;
+        const int luma_den = 10000;
```
Comment (wenbinc-Bin) [inline libavcodec/qsvdec.c]: According to MSDK doc. I think this should be 100000.

File: libavcodec/qsvdec.c [hunk 2]
```diff
@@ -628,6 +645,54 @@ static int qsv_export_film_grain(AVCodecContext *avctx, mfxExtAV1FilmGrainParam
...
+        int i;
+
+        if (!mastering)
+            return AVERROR(ENOMEM);
+
+        for (i = 0; i < 3; i++) {
+            const int j = mapping[i];
+            mastering->display_primaries[i][0] = av_make_q(mdcv->DisplayPrimariesX[j], chroma_den);
+            mastering->display_primaries[i][1] = av_make_q(mdcv->DisplayPrimariesY[j], chroma_den);
+            mastering->display_primaries[i][1].den = chroma_den;
```
Comment (wenbinc-Bin) [inline libavcodec/qsvdec.c]: I think this line is redundant.

File: libavcodec/qsvenc_hevc.c [hunk 3]
```diff
@@ -160,6 +161,82 @@ static int generate_fake_vps(QSVEncContext *q, AVCodecContext *avctx)
...
+            av_frame_get_side_data(frame, AV_FRAME_DATA_MASTERING_DISPLAY_METADATA);
+
+        if (sd) {
+            AVMasteringDisplayMetadata *mdm = (AVMasteringDisplayMetadata *)sd->data;
+
+            // SEI is needed when both the primaries and luminance are set
+            if (mdm->has_primaries && mdm->has_luminance) {
+                const int mapping[3] = {1, 2, 0};
+                const int chroma_den = 50000;
+                const int luma_den   = 10000;
```
Comment (wenbinc-Bin) [inline libavcodec/qsvenc_hevc.c]: I think this should be 100000

File: libavcodec/qsvenc_hevc.c [hunk 4]
```diff
@@ -160,6 +161,82 @@ static int generate_fake_vps(QSVEncContext *q, AVCodecContext *avctx)
...
+
+                mdcv->WhitePointX =
+                    FFMIN(lrint(chroma_den * av_q2d(mdm->white_point[0])),
+                          chroma_den);
+                mdcv->WhitePointY =
+                    FFMIN(lrint(chroma_den * av_q2d(mdm->white_point[1])),
+                          chroma_den);
+
+                mdcv->MaxDisplayMasteringLuminance =
+                    lrint(luma_den * av_q2d(mdm->max_luminance));
```
Comment (wenbinc-Bin) [inline libavcodec/qsvenc_hevc.c]: Do we need to clip MaxLuminance to 65535?

File: libavcodec/qsvdec.c (hunk 1 above)
Comment (xhaihao) [reply libavcodec/qsvdec.c]: I guess there is a typo in MSDK doc.  It is 0.0001 in the spec, and MaxDisplayMasteringLuminance and MinDisplayMasteringLuminance are read from SEI without conversion.

File: libavcodec/qsvdec.c (hunk 2 above)
Comment (xhaihao) [reply libavcodec/qsvdec.c]: Thanks for pointing it out, I'll delete it.

File: libavcodec/qsvenc_hevc.c (hunk 3 above)
Comment (xhaihao) [reply libavcodec/qsvenc_hevc.c]: See the above comment.

File: libavcodec/qsvenc_hevc.c (hunk 4 above)
Comment (xhaihao) [reply libavcodec/qsvenc_hevc.c]: According to the spec, this value can be greater than 65535.

Comment (xhaihao) [reply libavcodec/qsvenc_hevc.c]: the description in MSDK doc should be wrong.

File: libavcodec/qsvdec.c (hunk 1 above)
Comment (wenbinc-Bin) [reply libavcodec/qsvdec.c]: Ok, Got it.

File: libavcodec/qsvenc_hevc.c (hunk 3 above)
Comment (wenbinc-Bin) [reply libavcodec/qsvenc_hevc.c]: Got it.

File: libavcodec/qsvenc_hevc.c (hunk 4 above)
Comment (wenbinc-Bin) [reply libavcodec/qsvenc_hevc.c]: Ok, Got it. Thank you.

**Comment by wenbinc-Bin** [comment]
LGTM. Thanks

**Comment by uartie** [comment]
@xhaihao 3 failed 720p sharpen cases failing because it seems 50, 75, and 100 sharpen intensities all produce the same result...

We would expect PSNR to drop as sharpen intensity increases.  But all 3 cases report the same result

[41.846, 100.0, 100.0, 45.9514, 100.0, 100.0]

...this should not happen

**Comment by xhaihao** [comment]
@uartie Thanks for the finding, I'll take a look and update the patchset.

File: libavfilter/vf_vpp_qsv.c [hunk 5]
```diff
@@ -518,6 +526,109 @@ static int config_output(AVFilterLink *outlink)
     return 0;
 }
 
+static int update_frame_params(AVFilterContext *ctx, AVFrame *in)
+{
+#if QSV_ONEVPL
```
Comment (FurongZhang) [inline libavfilter/vf_vpp_qsv.c]: Just for my education purpose, this Marco QSV_ONEVPL is allowed in the code of FFMPEG? Thanks in advance!

Comment (xhaihao) [reply libavfilter/vf_vpp_qsv.c]: right.

File: libavfilter/vf_vpp_qsv.c [hunk 6]
```diff
@@ -518,6 +526,109 @@ static int config_output(AVFilterLink *outlink)
...
+
+            mdcv_conf.WhitePointX =
+                FFMIN(lrint(chroma_den * av_q2d(mdm->white_point[0])),
+                      chroma_den);
+            mdcv_conf.WhitePointY =
+                FFMIN(lrint(chroma_den * av_q2d(mdm->white_point[1])),
+                      chroma_den);
+
+            mdcv_conf.MaxDisplayMasteringLuminance =
+                lrint(luma_den * av_q2d(mdm->max_luminance));
```
Comment (FurongZhang) [inline libavfilter/vf_vpp_qsv.c]: Here, in VPL VPP definition, no need to multiple luma_den. It is in 1 nit. Not in luma_den.

Comment (xhaihao) [reply libavfilter/vf_vpp_qsv.c]: Here mdm->max_luminance = MaxDisplayMasteringLuminance / 10000,

Comment (FurongZhang) [reply libavfilter/vf_vpp_qsv.c]: ok, thanks for the explanation.

**Review by FurongZhang — DISMISSED** [review]

**Comment by nyanmisaka** [comment]
@xhaihao Given that `vf_tonemap_vaapi` is available on Gen 9.5 Kaby Lake and newer, I'm wondering if the QSV HDR-to-SDR tone-mapping supports Windows? Or is it still Linux only? Thanks!

Update: It does work on Gen12 platform but not on Gen9.5. Tried on Windows.

And this patch seems to break the `scale_mode=compute` on DG2 A380 since it always use SFC path instead of Compute, which is slow on dGPU: 100~120fps when downscaling from 4k HEVC to 1080p. I disabled the tonemap `tonemap=0` but it doesn't help.

image

**Comment by xhaihao** [comment]
> @xhaihao Given that `vf_tonemap_vaapi` is available on Gen 9.5 Kaby Lake and newer, I'm wondering if the QSV HDR-to-SDR tone-mapping supports Windows? Or is it still Linux only? Thanks!
> 
> Update: It does work on Gen12 platform but not on Gen9.5. Tried on Windows.

Thanks for trying it  on Windows.  Yes this feature is available for Gen12+ platforms only because the added ext buffers  (e.g. MFX_EXTBUFF_MASTERING_DISPLAY_COLOUR_VOLUME_IN) are not supported on legacy platforms).
> 
> And this patch seems to break the `scale_mode=compute` on DG2 A380 since it always use SFC path instead of Compute, which is slow on dGPU: 100~120fps when downscaling from 4k HEVC to 1080p. I disabled the tonemap `tonemap=0` but it doesn't help.
> 
> image

Thanks for raising this issue, I'll look into it.

**Comment by xhaihao** [comment]
The failures in pre-check were caused by the oneVPL GPU RT, I submitted https://github.com/oneapi-src/oneVPL-intel-gpu/pull/264 to fix it.

**Comment by nyanmisaka** [comment]
Updated the driver to [31.0.101.4146](https://www.intel.com/content/www/us/en/download/726609/intel-arc-iris-xe-graphics-whql-windows.html) and the my above issue persists.

Found some new issues:
- 4:2:2 and 4:4:4 HDR input causes blank/green VPP output, here's the [clip (yuv444p12)](https://drive.google.com/file/d/10Zex_XqUnOXFIHctOUsCVh4s7hjw1cio/view?usp=share_link) and [source file](https://nyaa.si/view/1239953).
- If the input frame has no HDR metadata then VPP fails with even if `tonemap=0`:
`[Parsed_vpp_qsv_1 @ 00000214c11d5ec0] Error running VPP: invalid video parameters (-15)`

**Comment by xhaihao** [comment]
> Updated the driver to [31.0.101.4146](https://www.intel.com/content/www/us/en/download/726609/intel-arc-iris-xe-graphics-whql-windows.html) and the my above issue persists.

QSV HDR->SDR works on gen12+ only. 

Do you mean scale_mode=compute:tonemap=0 still use SFC path ? 

> 
> Found some new issues:
> 
>     * 4:2:2 and 4:4:4 HDR input causes blank/green VPP output, here's the [clip (yuv444p12)](https://drive.google.com/file/d/10Zex_XqUnOXFIHctOUsCVh4s7hjw1cio/view?usp=share_link) and [source file](https://nyaa.si/view/1239953).

Thanks, I will look into this. 

> 
>     * If the input frame has no HDR metadata then VPP fails with even if `tonemap=0`:
>       `[Parsed_vpp_qsv_1 @ 00000214c11d5ec0] Error running VPP: invalid video parameters (-15)`

Is scale_mode set to compute  ?

**Comment by xhaihao** [comment]
@nyanmisaka  I failed to access your source file

**Comment by nyanmisaka** [comment]
arca380

`scale_mode=3` in ffmpeg master/6.0 build will always use compute mode on ARC GPU.

But with this patch added, `tonemap=0:scale_mode=3` results in compute mode always disabled.

```
ffmpeg -hwaccel qsv -hwaccel_output_format qsv -i any_video.mp4 -an -sn -vf vpp_qsv=format=nv12:tonemap=0:scale_mode=3 -c:v hevc_qsv -global_quality 22 -y out.mp4
```

> Is scale_mode set to compute ?

`scale_mode` doesn't affect this. The -15 error will be triggered if the decoded frame has no HDR metadata if `tonemap=1`.

> I failed to access your source file

The source video is quite large so I have to split it into parts with WINRAR so as they can be uploaded here.
[yuv444p12.part1.rar.zip](https://github.com/intel-media-ci/ffmpeg/files/10896636/yuv444p12.part1.rar.zip)
[yuv444p12.part2.rar.zip](https://github.com/intel-media-ci/ffmpeg/files/10896642/yuv444p12.part2.rar.zip)

**Comment by xhaihao** [comment]
> `scale_mode=3` in ffmpeg master/6.0 build will always use compute mode on ARC GPU.
> 
> But with this patch added, `tonemap=0:scale_mode=3` results in compute mode always disabled.

Seems it has different behaviors on Linux. 

I got the following error with a 10bit video

```
$ ffmpeg -hwaccel qsv -hwaccel_output_format qsv -i 10bit.mp4 -an -sn -vf vpp_qsv=format=nv12:scale_mode=3:tonemap=0 -c:v hevc_qsv -g  22 -y out.mp4
...
[Parsed_vpp_qsv_0 @ 0x55a0b0a28200] Error running VPP: device failed (-17)
Error while filtering: Input/output error
Failed to inject frame into filter network: Input/output error
Error while processing the decoded data for stream #0:0
```
The command below works well with a 8bit video and the compute engine is used
```
$ ffmpeg -hwaccel qsv -hwaccel_output_format qsv -i 8bit.mp4 -an -sn -vf vpp_qsv=format=p010:scale_mode=3:tonemap=0 -c:v hevc_qsv -g  22 -y out.mp4
```

> > Is scale_mode set to compute ?
> 
> `scale_mode` doesn't affect this. The -15 error will be triggered if the decoded frame has no HDR metadata if `tonemap=1`.
> 

I can't reproduce this on Linux. 
On windows,  I can reproduce this with 8bit videos but the command works well with 10bit videos.

**Comment by xhaihao** [comment]
@nyanmisaka 

> Found some new issues:
> 
>     * 4:2:2 and 4:4:4 HDR input causes blank/green VPP output, here's the [clip (yuv444p12)](https://drive.google.com/file/d/10Zex_XqUnOXFIHctOUsCVh4s7hjw1cio/view?usp=share_link) and [source file](https://nyaa.si/view/1239953).
> 
Could you provide a screenshot ? I want to make sure what we are seeing are the same.

**Comment by nyanmisaka** [comment]
> Could you provide a screenshot ? I want to make sure what we are seeing are the same.

AV_PIX_FMT_NV12 | YUV420P => Blank


AV_PIX_FMT_P010 | YUV420P10 => OK


AV_PIX_FMT_YUYV422 | YUV422P
AV_PIX_FMT_VUYX | YUV444P => Green


AV_PIX_FMT_Y210 | YUV422P10
AV_PIX_FMT_Y212 | YUV422P12
AV_PIX_FMT_XV30 | YUV444P10
AV_PIX_FMT_XV36 | YUV444P12
VPP filter produces serval frames but hangs the ffmpeg at exit.
image

I haven't checked these on Linux, results may vary.
AV_PIX_FMT_P010 | YUV420P10 is the only working pixel format when `tonemap=1` and the HDR metadata is available.

**Comment by xhaihao** [comment]
@nyanmisaka  Sorry for slow response, could you update your driver to [31.0.101.4502](https://www.intel.com/content/www/us/en/download/726609/intel-arc-iris-xe-graphics-whql-windows.html) ? 

> And this patch seems to break the scale_mode=compute on DG2 A380 since it always use SFC path instead of Compute
> If the input frame has no HDR metadata then VPP fails with even if tonemap=0:

The above 2 issues should be fixed in the latest patchset and gfx driver. 

> 4:2:2 and 4:4:4 HDR input causes blank/green VPP output

It may work on TGL etc but fail on  DG2 etc. It should be driver related.

**Comment by nyanmisaka** [comment]
> > And this patch seems to break the scale_mode=compute on DG2 A380 since it always use SFC path instead of Compute
> > If the input frame has no HDR metadata then VPP fails with even if tonemap=0:
> 
> The above 2 issues should be fixed in the latest patchset and gfx driver.

I can have a try tomorrow.

> > 4:2:2 and 4:4:4 HDR input causes blank/green VPP output
> 
> It may work on TGL etc but fail on DG2 etc. It should be driver related.

From what I understand, TGL+ platforms use LUT hardware to handle the HDR2SDR tone-mapping. Is there a restriction inside the driver or hardware that only allows P010 input? If so ffmpeg should prompt for this.

On Linux when I feed 4:4:4(xv30) or 4:2:2(y210) formats to `tonemap_vaapi`. It just crashes the driver with some errors.
```
/usr/include/c++/13.1.1/bits/stl_vector.h:1125: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = VP_SURFACE*; _Alloc = std::allocator<VP_SURFACE*>; reference = VP_SURFACE*&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
zsh: IOT instruction (core dumped)
```

**Comment by xhaihao** [comment]
> > > 4:2:2 and 4:4:4 HDR input causes blank/green VPP output
> > 
> > 
> > It may work on TGL etc but fail on DG2 etc. It should be driver related.
> 
> From what I understand, TGL+ platforms use LUT hardware to handle the HDR2SDR tone-mapping. Is there a restriction inside the driver or hardware that only allows P010 input? If so ffmpeg should prompt for this.
> 

AFAIK it can support other formats. 

> On Linux when I feed 4:4:4(xv30) or 4:2:2(y210) formats to `tonemap_vaapi`. It just crashes the driver with some errors.
> 
> ```
> /usr/include/c++/13.1.1/bits/stl_vector.h:1125: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = VP_SURFACE*; _Alloc = std::allocator<VP_SURFACE*>; reference = VP_SURFACE*&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
> zsh: IOT instruction (core dumped)
> ```

Did you use the latest media driver ? tonemap_vaapi works well with xv30 / y210 me. (Please apply https://github.com/intel-media-ci/ffmpeg/pull/662 if you are using the latest FFmpeg master)

$ ffmpeg -y -hwaccel vaapi -hwaccel_output_format vaapi -i 444_8b.mp4 -vf "tonemap_vaapi=format=nv12" -c:v hevc_vaapi out.h265
$ ffmpeg -y -hwaccel vaapi -hwaccel_output_format vaapi -i 422_8b.mp4 -vf "tonemap_vaapi=format=nv12" -c:v hevc_vaapi out.h265

**Comment by nyanmisaka** [comment]
Now I updated the patch set and the driver 4502 on DG2. These commands and the speed issue are fixed.
```
4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=compute:tonemap=1:format=nv12`
4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=low_power:tonemap=1:format=nv12`

frame= 2000 fps=356 q=-0.0 Lsize=N/A time=00:00:33.34 bitrate=N/A speed=5.93x
```

But new issues arises - FFmpeg **hang** if tonemap=1 and scale_mode=hq or auto.
```
4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=auto:tonemap=1:format=nv12`
4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=hq:tonemap=1:format=nv12`
4k p010 input => `vpp_qsv=scale_mode=auto:tonemap=1:format=nv12`
4k p010 input => `vpp_qsv=scale_mode=hq:tonemap=1:format=nv12`

frame=  124 fps=0.0 q=-0.0 size=N/A time=00:00:02.06 bitrate=N/A speed=3.64x     0x
```

And adding `out_color_primaries=bt709` will lead to device fail -17, so I have to leave it blank and only set `out_color_matrix=bt709:out_color_transfer=bt709`
```
4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=compute:tonemap=1:format=nv12:out_color_matrix=bt709:out_color_transfer=bt709:out_color_primaries=bt709`

[Parsed_vpp_qsv_0 @ 000001960a83ff80] Error running VPP: device failed (-17)
```

As for the 4:4:4/4:2:2 input and output. It will just fail.
```
4k p010 input => `vpp_qsv=w=1920:h=1080:tonemap=1:scale_mode=compute:out_color_matrix=bt709:out_color_transfer=bt709:format=xv36`
4k xv36 input => `vpp_qsv=w=1920:h=1080:tonemap=1:scale_mode=compute:out_color_matrix=bt709:out_color_transfer=bt709:format=nv12`

[Parsed_vpp_qsv_0 @ 000001c1ceea43c0] Error running VPP: device failed (-17)
```

**Comment by nyanmisaka** [comment]
> Did you use the latest media driver ? tonemap_vaapi works well with xv30 / y210 me. (Please apply #662 if you are using the latest FFmpeg master)

Sure. I'm using the latest media-driver on DG2 https://github.com/intel/media-driver/commit/0f3c366a3c2fa25649a8ab4885948c59057bd3b6

#662 doesn't help. `[with _Tp = VP_SURFACE*; _Alloc = std::allocator<VP_SURFACE*>; reference = VP_SURFACE*&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.` is an assertion inside of the media-driver.

Maybe it is platform specific? (DG2 fails but TGLx works) On DG2 it only allows P010 input.

**Comment by xhaihao** [comment]
> > Did you use the latest media driver ? tonemap_vaapi works well with xv30 / y210 me. (Please apply #662 if you are using the latest FFmpeg master)
> 
> Sure. I'm using the latest media-driver on DG2 [intel/media-driver@0f3c366](https://github.com/intel/media-driver/commit/0f3c366a3c2fa25649a8ab4885948c59057bd3b6)
> 
> #662 doesn't help.

#662 is to fix a FFmpeg regression when hw based filters are used

[vf#0:0 @ 0x5603ad5a3380] Error while filtering: Cannot allocate memory=  -0.0kbits/s speed=N/A    
Failed to inject frame into filter network: Cannot allocate memory
Cannot allocate memory
 
>  `[with _Tp = VP_SURFACE*; _Alloc = std::allocator<VP_SURFACE*>; reference = VP_SURFACE*&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.` is an assertion inside of the media-driver.
> 

Yes, it is an assertion in the media driver but I can't reproduce this assertion, hence I thought this assertion has gone away if using the latest driver. 

> Maybe it is platform specific? (DG2 fails but TGLx works) On DG2 it only allows P010 input.

It works for me on DG2.  What are the building options for your media driver ? 

If possible, could you share your 444 / 422 8bit clips ?

**Comment by xhaihao** [comment]
> Now I updated the patch set and the driver 4502 on DG2. These commands and the speed issue are fixed.
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=compute:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=low_power:tonemap=1:format=nv12`
> 
> frame= 2000 fps=356 q=-0.0 Lsize=N/A time=00:00:33.34 bitrate=N/A speed=5.93x
> ```
> 
> But new issues arises - FFmpeg **hang** if tonemap=1 and scale_mode=hq or auto.
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=auto:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=hq:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=scale_mode=auto:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=scale_mode=hq:tonemap=1:format=nv12`
> 
> frame=  124 fps=0.0 q=-0.0 size=N/A time=00:00:02.06 bitrate=N/A speed=3.64x     0x
> ```
> 
> And adding `out_color_primaries=bt709` will lead to device fail -17, so I have to leave it blank and only set `out_color_matrix=bt709:out_color_transfer=bt709`
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=compute:tonemap=1:format=nv12:out_color_matrix=bt709:out_color_transfer=bt709:out_color_primaries=bt709`
> 
> [Parsed_vpp_qsv_0 @ 000001960a83ff80] Error running VPP: device failed (-17)
> ```
> 
> As for the 4:4:4/4:2:2 input and output. It will just fail.
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:tonemap=1:scale_mode=compute:out_color_matrix=bt709:out_color_transfer=bt709:format=xv36`
> 4k xv36 input => `vpp_qsv=w=1920:h=1080:tonemap=1:scale_mode=compute:out_color_matrix=bt709:out_color_transfer=bt709:format=nv12`
> 
> [Parsed_vpp_qsv_0 @ 000001c1ceea43c0] Error running VPP: device failed (-17)
> ```

Thanks for testing the new patchset & driver. Seems oneVPL or media driver doesn't support some combinations of tonemapping and other operations. How about adding a separate tonemap_qsv filter, like as tonemap_vaapi ?

**Comment by nyanmisaka** [comment]
> #662 is to fix a FFmpeg regression when hw based filters are used
> 
> [vf#0:0 @ 0x5603ad5a3380] Error while filtering: Cannot allocate memory= -0.0kbits/s speed=N/A Failed to inject frame into filter network: Cannot allocate memory Cannot allocate memory

I have observed this issue many times. It happens when using VAAPI/QSV filters with encoder settings `-look_ahead_depth` or `-bf`. Increasing the `extra_hw_frames` can "fix" it but this will take more VRAM.

> If possible, could you share your 444 / 422 8bit clips ?

- [yuyv422.zip](https://github.com/intel-media-ci/ffmpeg/files/11879232/yuyv422.zip)
- [vuyx.zip](https://github.com/intel-media-ci/ffmpeg/files/11879235/vuyx.zip)

Generally, HDR videos do not use 8bit encoding.

**Comment by nyanmisaka** [comment]
> Thanks for testing the new patchset & driver. Seems oneVPL or media driver doesn't support some combinations of tonemapping and other operations. How about adding a separate tonemap_qsv filter, like as tonemap_vaapi ?

I agree with separating `tonemap_qsv` from `vpp_qsv` if the driver cannot do all VPP operations in one call.

**Comment by xhaihao** [comment]
> Now I updated the patch set and the driver 4502 on DG2. These commands and the speed issue are fixed.
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=compute:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=low_power:tonemap=1:format=nv12`
> 
> frame= 2000 fps=356 q=-0.0 Lsize=N/A time=00:00:33.34 bitrate=N/A speed=5.93x
> ```
> 
> But new issues arises - FFmpeg **hang** if tonemap=1 and scale_mode=hq or auto.
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=auto:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=hq:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=scale_mode=auto:tonemap=1:format=nv12`
> 4k p010 input => `vpp_qsv=scale_mode=hq:tonemap=1:format=nv12`
> 
> frame=  124 fps=0.0 q=-0.0 size=N/A time=00:00:02.06 bitrate=N/A speed=3.64x     0x
> ```


@nyanmisaka  It works well for me on DG2 / TGL + Windows / Linux, could you provide your command and test clip ?  


> And adding `out_color_primaries=bt709` will lead to device fail -17, so I have to leave it blank and only set `out_color_matrix=bt709:out_color_transfer=bt709`
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:scale_mode=compute:tonemap=1:format=nv12:out_color_matrix=bt709:out_color_transfer=bt709:out_color_primaries=bt709`
> 
> [Parsed_vpp_qsv_0 @ 000001960a83ff80] Error running VPP: device failed (-17)
> ```

It also works well for me,  what's your command and test clip ?  


> As for the 4:4:4/4:2:2 input and output. It will just fail.
> 
> ```
> 4k p010 input => `vpp_qsv=w=1920:h=1080:tonemap=1:scale_mode=compute:out_color_matrix=bt709:out_color_transfer=bt709:format=xv36`
> 4k xv36 input => `vpp_qsv=w=1920:h=1080:tonemap=1:scale_mode=compute:out_color_matrix=bt709:out_color_transfer=bt709:format=nv12`
> 
> [Parsed_vpp_qsv_0 @ 000001c1ceea43c0] Error running VPP: device failed (-17)
> ```

4K p010 input + xv36/xv30/y212/y210 output works well for me. 

4K xv36/xv30/y212/y210 input works well on TGL + Linux / Windows, but doesn't work on DG2 + Linux / Windows. We identified a regression in the gfx driver, and will fix it in the next release.

**Comment by xhaihao** [comment]
> > #662 is to fix a FFmpeg regression when hw based filters are used
> > [vf#0:0 @ 0x5603ad5a3380] Error while filtering: Cannot allocate memory= -0.0kbits/s speed=N/A Failed to inject frame into filter network: Cannot allocate memory Cannot allocate memory
> 
> I have observed this issue many times. It happens when using VAAPI/QSV filters with encoder settings `-look_ahead_depth` or `-bf`. Increasing the `extra_hw_frames` can "fix" it but this will take more VRAM.

This will be fixed in https://github.com/intel-media-ci/ffmpeg/pull/635

> 
> > If possible, could you share your 444 / 422 8bit clips ?
> 
>     * [yuyv422.zip](https://github.com/intel-media-ci/ffmpeg/files/11879232/yuyv422.zip)
> 
>     * [vuyx.zip](https://github.com/intel-media-ci/ffmpeg/files/11879235/vuyx.zip)
> 
> 

Thanks for sharing your clips.

**Comment by nyanmisaka** [comment]
> It works well for me on DG2 / TGL + Windows / Linux, could you provide your command and test clip ?

> It also works well for me, what's your command and test clip ?

@xhaihao 
These two only happens with some HEVC clips.
For example [sony_clip_p010.zip](https://github.com/intel-media-ci/ffmpeg/files/11965968/sony_clip_p010.zip) cut down from https://4kmedia.org/sony-camping-in-nature-4k-demo/
It is guessed that it may be related to HDR metadata since some clips are not affected.

> 4K p010 input + xv36/xv30/y212/y210 output works well for me.

These do work when the input file does not contain any HDR metadata.
Using the above clip will be able to reproduce the -17 error.

> 4K xv36/xv30/y212/y210 input works well on TGL + Linux / Windows, but doesn't work on DG2 + Linux / Windows. We identified a regression in the gfx driver, and will fix it in the next release.

That's good to know.

**Comment by xhaihao** [comment]
@nyanmisaka  I failed to reproduce the hang issue but may reproduce the -17 error with your clip.  It seems FFmpeg insert extradata in a wrong position, you may try with https://patchwork.ffmpeg.org/project/ffmpeg/patch/20220317063546.7429-1-haihao.xiang@intel.com/

**Comment by xhaihao** [comment]
merged

**Comment by FurongZhang** [comment]
Thank you so much Haihao.

**Pull request closed without merging**